File: windows/companion/src/PicoSwitch.Companion.Services/Presentation/AmiiboInspection.cs
using PicoSwitch.Management;

namespace PicoSwitch.Companion.Services.Presentation;

/// <summary>One label/value line in the inspection pane.</summary>
/// <param name="Monospace">
/// For identifiers a person compares character by character — a UID or a figure
/// id — where proportional digits make scanning harder.
/// </param>
public sealed record AmiiboDetailRow(string Label, string Value, bool Monospace = false);

/// <summary>A named group of related lines.</summary>
public sealed record AmiiboDetailGroup(string Title, ValueList<AmiiboDetailRow> Rows)
{
    public bool Any => Rows.Count > 0;
}

/// <summary>
/// The selected Amiibo, organised the way someone actually reads it.
/// </summary>
/// <remarks>
/// REPLACES A FLAT KEY/VALUE SLAB. The same facts were all present before, in
/// one undifferentiated column, which made the page read like a database row
/// rather than a description of a figure. Grouping them means a user can find
/// "who owns this" without reading past the tag size.
///
/// Groups are omitted when they have nothing to say. A tag that was never set up
/// has no owner, no nickname and no dates, and printing five "—" lines is worse
/// than printing none: it implies the information exists and is missing, when in
/// fact the tag is simply blank.
///
/// Pure, so the grouping and omission rules are tested without a page.
/// </remarks>
public static class AmiiboInspection
{
    public static IReadOnlyList<AmiiboDetailGroup> Build(
        AmiiboLibraryItem item,
        AmiiboCatalogEntry? catalog,
        AmiiboDetails? decoded,
        bool onAdapter,
        bool adapterChanged)
    {
        var groups = new List<AmiiboDetailGroup>
        {
            Group("Identity", Identity(item, catalog)),
            Group("Tag", Tag(item, decoded)),
        };

        // Only when the tag carries them. See the remarks above.
        var registration = Registration(decoded);
        if (registration.Count > 0)
        {
            groups.Add(Group("Registration", registration));
        }

        groups.Add(Group("Game data", GameData(decoded, catalog)));
        groups.Add(Group("Adapter", Adapter(onAdapter, adapterChanged)));

        return [.. groups.Where(group => group.Any)];
    }

    private static AmiiboDetailGroup Group(string title, List<AmiiboDetailRow> rows) =>
        new(title, new ValueList<AmiiboDetailRow>(rows));

    private static List<AmiiboDetailRow> Identity(
        AmiiboLibraryItem item, AmiiboCatalogEntry? catalog)
    {
        var rows = new List<AmiiboDetailRow> { new("Your name", item.DisplayName) };

        var figure = catalog?.Name is { Length: > 0 } name
            ? name
            : catalog?.Character is { Length: > 0 } character
                ? character
                : null;

        // Only when it says something the user's own name does not.
        if (figure is not null &&
            !string.Equals(figure, item.DisplayName, StringComparison.CurrentCultureIgnoreCase))
        {
            rows.Add(new AmiiboDetailRow("Figure", figure));
        }

        Add(rows, "Game series", catalog?.GameSeries);
        Add(rows, "Collection", catalog?.AmiiboSeries);
        Add(rows, "Type", catalog?.Type);
        Add(rows, "Released", catalog?.ReleaseDate);
        rows.Add(new AmiiboDetailRow("Figure ID", item.FigureId, Monospace: true));
        return rows;
    }

    private static List<AmiiboDetailRow> Tag(
        AmiiboLibraryItem item, AmiiboDetails? decoded) =>
    [
        new("Tag type", item.TagType == AmiiboTagType.FigureV3 ? "Figure v3" : "NTAG215"),
        new("UID", item.Uid, Monospace: true),
        new("Size", $"{item.Size} bytes"),
        new("Contents", Contents(decoded)),
    ];

    /// <summary>
    /// What could be read from the encrypted body, in the user's terms.
    /// </summary>
    /// <remarks>
    /// Three distinguishable answers, because they need three different actions:
    /// import a key, investigate a bad file, or nothing at all.
    /// </remarks>
    private static string Contents(AmiiboDetails? decoded) => decoded?.Crypto switch
    {
        AmiiboCryptoState.Valid => decoded.SetUp ? "Set up" : "Blank, never set up",
        AmiiboCryptoState.Invalid => "Could not be decrypted with the imported keys",
        _ => "Import your Amiibo keys to read this",
    };

    private static List<AmiiboDetailRow> Registration(AmiiboDetails? decoded)
    {
        var rows = new List<AmiiboDetailRow>();
        if (decoded?.Crypto != AmiiboCryptoState.Valid || !decoded.SetUp)
        {
            return rows;
        }

        Add(rows, "Nickname", decoded.Nickname);
        Add(rows, "Owner", decoded.Owner);
        Add(rows, "Registered", decoded.SetupDate);
        Add(rows, "Last written", decoded.LastWriteDate);
        if (decoded.WriteCounter > 0)
        {
            rows.Add(new AmiiboDetailRow("Times written", decoded.WriteCounter.ToString()));
        }

        return rows;
    }

    private static List<AmiiboDetailRow> GameData(
        AmiiboDetails? decoded, AmiiboCatalogEntry? catalog)
    {
        if (decoded?.Crypto != AmiiboCryptoState.Valid)
        {
            return [];
        }

        if (!decoded.HasAppData)
        {
            // "Game", not "Game data": the group heading already says that, and
            // the row label must match the has-data branch below so the two
            // states line up in the same column.
            return [new AmiiboDetailRow("Game", "None")];
        }

        // The catalog can name the game where the firmware's small built-in
        // table cannot; its answer is preferred when it has one.
        var named = decoded.TitleId.Length > 0
            ? catalog?.TitleIds.FirstOrDefault(pair =>
                string.Equals(pair.Key, decoded.TitleId, StringComparison.OrdinalIgnoreCase)).Value
            : null;

        var rows = new List<AmiiboDetailRow>
        {
            new("Game", named is { Length: > 0 } ? named : decoded.AppDataLabel),
        };

        if (decoded.TitleId.Length > 0)
        {
            rows.Add(new AmiiboDetailRow("Title ID", decoded.TitleId, Monospace: true));
        }

        return rows;
    }

    private static List<AmiiboDetailRow> Adapter(bool onAdapter, bool changed) =>
    [
        new("Status", changed
            ? "On the adapter, changed by the console"
            : onAdapter
                ? "On the adapter"
                : "Not on the adapter"),
    ];

    private static void Add(List<AmiiboDetailRow> rows, string label, string? value)
    {
        if (value is { Length: > 0 })
        {
            rows.Add(new AmiiboDetailRow(label, value));
        }
    }
}

File: android/companion/app/src/main/java/dev/picoswitch/companion/data/AmiiboInspection.kt
package dev.picoswitch.companion.data

import dev.picoswitch.companion.model.AmiiboCatalogEntry
import dev.picoswitch.companion.model.AmiiboCryptoState
import dev.picoswitch.companion.model.AmiiboDetails
import dev.picoswitch.companion.model.AmiiboLibraryItem
import dev.picoswitch.companion.model.AmiiboTagType

/**
 * Which part of the selected Amiibo is being inspected.
 *
 * The same three the Windows companion uses. Only one category's rows are built
 * at a time — rendering all of them into one column is what turns an inspector
 * into a scrolling settings form.
 */
enum class AmiiboCategory { Overview, Tag, Adapter }

/**
 * One label/value line.
 *
 * @param monospace for identifiers a person compares character by character — a
 * UID or a figure id — where proportional digits make scanning harder.
 */
data class AmiiboDetailRow(
    val label: String,
    val value: String,
    val monospace: Boolean = false,
)

/** A named group of related lines. */
data class AmiiboDetailGroup(val title: String, val rows: List<AmiiboDetailRow>) {
    val any: Boolean get() = rows.isNotEmpty()
}

/**
 * The selected Amiibo, organised the way someone actually reads it.
 *
 * A DELIBERATE MIRROR of the Windows `AmiiboInspection`, so the two companions
 * describe a figure with the same groups, the same labels and the same omission
 * rules. The UIs look nothing alike; what a user is told should not.
 *
 * Groups are omitted when they have nothing to say. A tag that was never set up
 * has no owner, no nickname and no dates, and printing five empty lines implies
 * the information exists and is missing when the tag is simply blank.
 *
 * Pure, so the grouping and omission rules are tested without Compose.
 */
object AmiiboInspection {

    fun build(
        item: AmiiboLibraryItem,
        catalog: AmiiboCatalogEntry?,
        decoded: AmiiboDetails?,
        onAdapter: Boolean,
        adapterChanged: Boolean,
    ): List<AmiiboDetailGroup> = listOfNotNull(
        AmiiboDetailGroup("Identity", identity(item, catalog)),
        AmiiboDetailGroup("Tag", tag(item, decoded)),
        registration(decoded).takeIf { it.isNotEmpty() }
            ?.let { AmiiboDetailGroup("Registration", it) },
        AmiiboDetailGroup("Game data", gameData(decoded, catalog)),
        AmiiboDetailGroup("Adapter", adapter(onAdapter, adapterChanged)),
    ).filter { it.any }

    /** The rows for one category, which is all the UI ever shows at once. */
    fun forCategory(
        category: AmiiboCategory,
        groups: List<AmiiboDetailGroup>,
    ): List<AmiiboDetailGroup> {
        val wanted = when (category) {
            AmiiboCategory.Overview -> setOf("Identity")
            AmiiboCategory.Tag -> setOf("Tag", "Registration", "Game data")
            AmiiboCategory.Adapter -> setOf("Adapter")
        }
        return groups.filter { it.title in wanted }
    }

    private fun identity(
        item: AmiiboLibraryItem,
        catalog: AmiiboCatalogEntry?,
    ): List<AmiiboDetailRow> = buildList {
        add(AmiiboDetailRow("Your name", item.displayName))

        val figure = catalog?.name?.takeIf { it.isNotBlank() }
            ?: catalog?.character?.takeIf { it.isNotBlank() }

        // Only when it says something the user's own name does not.
        if (figure != null && !figure.equals(item.displayName, ignoreCase = true)) {
            add(AmiiboDetailRow("Figure", figure))
        }

        catalog?.gameSeries?.takeIf { it.isNotBlank() }
            ?.let { add(AmiiboDetailRow("Game series", it)) }
        catalog?.amiiboSeries?.takeIf { it.isNotBlank() }
            ?.let { add(AmiiboDetailRow("Collection", it)) }
        catalog?.type?.takeIf { it.isNotBlank() }
            ?.let { add(AmiiboDetailRow("Type", it)) }
        catalog?.releaseDate?.takeIf { it.isNotBlank() }
            ?.let { add(AmiiboDetailRow("Released", it)) }

        add(AmiiboDetailRow("Figure ID", item.figureId, monospace = true))
    }

    private fun tag(
        item: AmiiboLibraryItem,
        decoded: AmiiboDetails?,
    ): List<AmiiboDetailRow> = listOf(
        AmiiboDetailRow(
            "Tag type",
            if (item.tagType == AmiiboTagType.FigureV3) "Figure v3" else "NTAG215",
        ),
        AmiiboDetailRow("UID", item.uid, monospace = true),
        AmiiboDetailRow("Size", "${item.size} bytes"),
        AmiiboDetailRow("Contents", contents(decoded)),
    )

    /**
     * What could be read from the encrypted body, in the user's terms.
     *
     * Three distinguishable answers, because they need three different actions:
     * import a key, investigate a bad file, or nothing at all.
     */
    private fun contents(decoded: AmiiboDetails?): String = when (decoded?.crypto) {
        AmiiboCryptoState.Valid -> if (decoded.setUp) "Set up" else "Blank, never set up"
        AmiiboCryptoState.Invalid -> "Could not be decrypted with the imported keys"
        else -> "Import your Amiibo keys to read this"
    }

    private fun registration(decoded: AmiiboDetails?): List<AmiiboDetailRow> {
        if (decoded?.crypto != AmiiboCryptoState.Valid || !decoded.setUp) return emptyList()
        return buildList {
            decoded.nickname.takeIf { it.isNotBlank() }
                ?.let { add(AmiiboDetailRow("Nickname", it)) }
            decoded.owner.takeIf { it.isNotBlank() }
                ?.let { add(AmiiboDetailRow("Owner", it)) }
            decoded.setupDate?.takeIf { it.isNotBlank() }
                ?.let { add(AmiiboDetailRow("Registered", it)) }
            decoded.lastWriteDate?.takeIf { it.isNotBlank() }
                ?.let { add(AmiiboDetailRow("Last written", it)) }
            decoded.writeCounter?.takeIf { it > 0 }
                ?.let { add(AmiiboDetailRow("Times written", it.toString())) }
        }
    }

    private fun gameData(
        decoded: AmiiboDetails?,
        catalog: AmiiboCatalogEntry?,
    ): List<AmiiboDetailRow> {
        if (decoded?.crypto != AmiiboCryptoState.Valid) return emptyList()
        if (decoded.hasAppData != true) return listOf(AmiiboDetailRow("Game", "None"))

        // The catalog can name the game the firmware's small built-in table
        // cannot; its answer is preferred when it has one.
        val named = decoded.titleId.takeIf { it.isNotBlank() }
            ?.let { titleId -> catalog?.titleIds?.entries?.firstOrNull { it.key.equals(titleId, true) }?.value }

        return buildList {
            add(AmiiboDetailRow("Game", named ?: decoded.appDataLabel))
            decoded.titleId.takeIf { it.isNotBlank() }
                ?.let { add(AmiiboDetailRow("Title ID", it, monospace = true)) }
        }
    }

    private fun adapter(onAdapter: Boolean, changed: Boolean): List<AmiiboDetailRow> = listOf(
        AmiiboDetailRow(
            "Status",
            when {
                changed -> "On the adapter, changed by the console"
                onAdapter -> "On the adapter"
                else -> "Not on the adapter"
            },
        ),
    )
}
